### PR TITLE
SONAR-7276 support multiple conditions for a given metric in QG

### PR DESCRIPTION
**Pull request by sns-seb**
they must have a different period

### CODE REVIEW

File: server/sonar-server/src/main/java/org/sonar/server/computation/step/QualityGateMeasuresStep.java [hunk 1]
```diff
@@ -245,4 +262,24 @@ public Condition apply(@Nonnull EvaluatedCondition input) {
...
+      this.condition = condition;
+    }
+  }
+
+  private enum ConditionToMetric implements Function<Condition, Metric> {
+    INSTANCE;
+
+    @Override
+    @Nonnull
+    public Metric apply(@Nonnull Condition input) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/computation/step/QualityGateMeasuresStep.java]: MAJOR Remove this "Nonnull" annotation to honor the overridden method's contract. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2638)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue
* MAJOR 1 major

Watch the comments in this conversation to review them.

File: server/sonar-server/src/main/java/org/sonar/server/computation/step/QualityGateMeasuresStep.java [hunk 2]
```diff
@@ -164,26 +166,39 @@ private void updateQualityGateStatusHolder(QualityGate qualityGate, QualityGateD
...
         .create();
-      measureRepository.update(project, condition.getMetric(), updatedMeasure);
+      measureRepository.update(project, metric, updatedMeasure);
 
-      builder.addEvaluatedCondition(condition, evaluationResult);
+      builder.addEvaluatedCondition(metricEvaluationResult);
     }
   }
 
+  private static MetricEvaluationResult evaluateQualityGate(Measure measure, Collection<Condition> conditions) {
```
Comment (simonbrandhof) [inline server/sonar-server/src/main/java/org/sonar/server/computation/step/QualityGateMeasuresStep.java]: the `conditions` should be sorted in order to have deterministic results when multiple conditions (max 2 in theory) fail with the same severity. I suggest that condition on leak period takes precedence on the absolute condition.

**Comment by sns-seb** [comment]
@simonbrandhof feedback applied

File: server/sonar-server/src/main/java/org/sonar/server/computation/step/QualityGateMeasuresStep.java [hunk 3]
```diff
@@ -245,4 +265,34 @@ public Condition apply(@Nonnull EvaluatedCondition input) {
...
+      return input.getMetric();
+    }
+  }
+
+  public enum ConditionToPeriod implements Function<Condition, Integer> {
+    INSTANCE;
+
+    @Override
+    @Nullable
+    public Integer apply(@Nonnull Condition input) {
```
Comment (SonarTech) [inline server/sonar-server/src/main/java/org/sonar/server/computation/step/QualityGateMeasuresStep.java]: MAJOR Remove this "Nonnull" annotation to honor the overridden method's contract. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2638)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* MAJOR 2 major

Watch the comments in this conversation to review them.

**Comment by simonbrandhof** [comment]
LGTM

**Pull request merged**